### DISCUSSION TITLE
Possible to specify ESP payload’s source interface?

### DISCUSSION THREAD

**Discussion by hgl** [strongswan/strongswan#851] [Q&A] [answered]
My roadwarrior client connects to my IKEv2 server WAN interface, and it gets its virtual IP from LAN’s DHCP server. When ESP is decrypted, the packets, although have a LAN ip as source IP, are forwarded from WAN to LAN. This forces my firewall rules to be unnecessarily complicated, in order to allow the roadwarrior.

I wonder if it’s possible to specify that after decrying ESP, packets should originate from the LAN interface?

**Comment by tobiasbrunner** [comment, marked as answer]
I don't think the Linux kernel allows this (except perhaps via XFRM interfaces, didn't you mention them before?). But what exactly is complicated?

**Reply by hgl** [reply]
With XFRM interface, the packets still can’t originate from the LAN interface I guess?

My filter’s forward chain has a default rejected policy, and I have a bunch rules that allow forwarding between the LAN and other  interfaces. So for the roadwarrior to work, I have to duplicate these forwarding rules.

May be I should identify LAN by ip range instead of interfaces?

**Reply by hgl** [reply]
But it’s not straightforward to identify IPv6 LAN with IP range I guess?

**Reply by tobiasbrunner** [reply]
> With XFRM interface, the packets still can’t originate from the LAN interface I guess?

Yes, but it's at least no the WAN interface so you could handle that more restrictively.

> So for the roadwarrior to work, I have to duplicate these forwarding rules.

You could use the policy module to match only decrypted packets. Actually, the default [updown](https://wiki.strongswan.org/projects/strongswan/wiki/updown) script can install such rules in the FORWARD and optionally the INPUT chains automatically (you can adapt a copy of the script if they are not specific enough).

> May be I should identify LAN by ip range instead of interfaces?

Possible, but then you should probably ensure that these ranges are blocked on the WAN interface for packets that were not decrypted.

> But it’s not straightforward to identify IPv6 LAN with IP range I guess?

What do you mean?

**Reply by hgl** [reply]
> Yes, but it's at least no the WAN interface so you could handle that more restrictively.

This makes sense, and I took your advice and created an XFRM interface for that. However, ARP requests asking for client's MAC address from the LAN interface never got any responses. I wonder if it's a limitation with the DHCP + XFRM combination (since the XFRM interface doesn't have a MAC address) or my firewall rules are at play again (I don't think I have any rules that can block ARPs)?

> You could use the policy module to match only decrypted packets.

Are you referring to [rules like this](https://github.com/strongswan/strongswan/blob/85d626e9aef9f12233fe79ea0b6fb61c89ff149d/src/_updown/_updown.in#L245)? Maybe I'm looking at the wrong one, but doesn't it just match the encrypted ESP or AH packets?

> blocked on the WAN interface for packets that were not decrypted.

Why? Shouldn't my ISP refuse to route such ESP packets with LAN IPs as destinations to my WAN interface? Or it's a safety measure against the ISP itself?

> What do you mean?

For IPv4, with the help of DHCP, I can be in total control of which interfaces/hosts get assigned the LAN IPs. But with IPv6, SLAAC is employed, and all interfaces assign themselves with `fe80::/10`. Since I only know what interfaces (and hosts behind them) belong to LAN in advance, I can't really tell what IPv6 addresses belong to LAN in advance?

**Reply by tobiasbrunner** [reply]
> However, ARP requests asking for client's MAC address from the LAN interface never got any responses.

[FARPPlugin](https://wiki.strongswan.org/projects/strongswan/wiki/FARPPlugin)



> Maybe I'm looking at the wrong one, but doesn't it just match the encrypted ESP or AH packets?

No, why do you think so? Because it's added to INPUT? That's because the `up-host:iptables` case is for host-to-host connections. For subnets you want to look at the rules in `up-client:iptables`.

> Shouldn't my ISP refuse to route such ESP packets with LAN IPs as destinations to my WAN interface?

It's not about ESP packets but about injected plaintext packets. And yes, it's a safeguard against attackers close to your server.



> Since I only know what interfaces (and hosts behind them) belong to LAN in advance, I can't really tell what IPv6 addresses belong to LAN in advance?

Those are link-local addresses, so it's definitely not a good idea to base firewall rules on them.

**Reply by hgl** [reply]
Thanks for the help.

I really like the XFRM interface idea, and I'd really like to get it to work.

> FARPPlugin

I do have that plugin loaded:
```
00[LIB] loaded plugins: charon nonce x509 pem openssl kernel-netlink socket-default farp vici dhcp
```
swanctl.conf

```
connections {
	home {
		version = 2
		pools = dhcp
		proposals = aes128gcm16-prfsha256-ecp256,aes128gcm16-prfsha256-x25519
		send_certreq = no
		childless = never
		if_id_in = 1
		if_id_out = 1
		local {
			auth = pubkey
		}
		remote {
			auth = pubkey
		}
		children {
			child {
				local_ts = 0.0.0.0/0
				esp_proposals = aes128gcm16-ecp256,aes128gcm16-x25519
			}
		}
	}
}
```
strongswan.conf
```
charon {
	keep_alive = 0
	filelog {
		stderr {
		}
	}
	start-scripts {
		swanctl = swanctl --load-all --noprompt
	}
	plugins {
		dhcp {
			force_server_address = 192.168.1.1
			server = 192.168.1.255
		}
	}
}
```

```
$ ip -d addr show vpn
vpn@pppoe-wan: <MULTICAST,NOARP,UP,LOWER_UP> mtu 1492 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none  promiscuity 2 minmtu 68 maxmtu 65535 
    xfrm if_id 0x1 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 
    inet6 <link-local-addr> scope link stable-privacy 
       valid_lft forever preferred_lft forever
```
The clients can successfully connect to the server and get an IP `192.168.1.235` via DHCP.
This is captured on the `vpn` interface filtered only with `host 192.168.1.235`
![image](https://user-images.githubusercontent.com/7038035/149494963-9181c232-6775-43aa-83a1-45cde7ac5317.png)

And this is captured on the LAN interface filtered only with `host 192.168.1.235`
![image](https://user-images.githubusercontent.com/7038035/149495980-bf72a752-9baf-4063-96b0-9a6ee722a783.png)

Any idea why ARPs are not responded?

> No, why do you think so?

Sorry, I thought the policy module only matches encrypted packets, looks like it actually matches decrypted ones.

**Reply by tobiasbrunner** [reply]
> Any idea why ARPs are not responded?

No, check the log (there should be at least one `listening for ARP requests ...` message, and on level 2 for _net_ there is one for every ARP reply).

**Reply by hgl** [reply]
> check the log

My log doesn't contain that line at all. I guess that means the farp plugin is loaded but not actually working? Is it possible to debug why it's not listening?

**Reply by hgl** [reply]
I don't see any `farp cannot find interface ...` lines either.

**Reply by hgl** [reply]
And here is the full log, captured after the client has connected to the IKEv2 VPN, but can't access network behind the server.

```
00[DMN] Starting IKE charon daemon (strongSwan 5.9.2, Linux 5.4.154, x86_64)
00[LIB] loaded plugins: charon nonce x509 pem openssl kernel-netlink socket-default farp vici dhcp
00[JOB] spawning 16 worker threads
00[DMN] executing start script 'swanctl' (swanctl --load-all --noprompt)
03[NET] waiting for data on sockets
05[CFG] loaded certificate 'CN=server'
00[DMN] swanctl: loaded certificate from '/etc/swanctl/x509/home.crt'
10[CFG] loaded certificate 'CN=CA'
14[CFG] loaded ECDSA private key
16[CFG] added vici connection: home
00[DMN] swanctl: loaded certificate from '/etc/swanctl/x509ca/home.crt'
00[DMN] swanctl: loaded ecdsa key from '/etc/swanctl/ecdsa/home.key'
00[DMN] swanctl: no authorities found, 0 unloaded
00[DMN] swanctl: no pools found, 0 unloaded
00[DMN] swanctl: loaded connection 'home'
00[DMN] swanctl: successfully loaded 1 connections, 0 unloaded
03[NET] received packet: from <client-ipv6>[500] to <server-ipv6>[500]
03[NET] waiting for data on sockets
08[NET] received packet: from <client-ipv6>[500] to <server-ipv6>[500] (232 bytes)
08[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(REDIR_SUP) N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) ]
08[IKE] <client-ipv6> is initiating an IKE_SA
08[CFG] selected proposal: IKE:AES_GCM_16_128/PRF_HMAC_SHA2_256/ECP_256
08[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(CHDLESS_SUP) N(MULT_AUTH) ]
08[NET] sending packet: from <server-ipv6>[500] to <client-ipv6>[500] (256 bytes)
02[NET] sending packet: from <server-ipv6>[500] to <client-ipv6>[500]
03[NET] received packet: from <client-ipv6>[4500] to <server-ipv6>[4500]
03[NET] waiting for data on sockets
14[NET] received packet: from <client-ipv6>[4500] to <server-ipv6>[4500] (800 bytes)
14[ENC] unknown attribute type INTERNAL_DNS_DOMAIN
14[ENC] parsed IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH CPRQ(ADDR SUBNET MASK DHCP DNS ADDR6 SUBNET6 DHCP6 DNS6 DOMAIN) N(ESP_TFC_PAD_N) N(NON_FIRST_FRAG) SA TSi TSr N(MOBIKE_SUP) ]
14[IKE] received cert request for "CN=CA"
14[IKE] received end entity cert "CN=client"
14[CFG] looking for peer configs matching <server-ipv6>[server]...<client-ipv6>[client]
14[CFG] selected peer config 'home'
14[CFG]   using certificate "CN=client"
14[CFG]   using trusted ca certificate "CN=CA"
14[CFG]   reached self-signed root ca with a path length of 0
14[IKE] authentication of 'client' with ECDSA-256 signature successful
14[IKE] received ESP_TFC_PADDING_NOT_SUPPORTED, not using ESPv3 TFC padding
14[IKE] peer supports MOBIKE
14[IKE] authentication of 'server' (myself) with ECDSA-256 signature successful
14[IKE] IKE_SA home[1] established between <server-ipv6>[server]...<client-ipv6>[client]
14[IKE] scheduling rekeying in 14265s
14[IKE] maximum IKE_SA lifetime 15705s
14[IKE] sending end entity cert "CN=server"
14[IKE] peer requested virtual IP %any
14[CFG] sending DHCP DISCOVER to 192.168.1.255
14[CFG] sending DHCP DISCOVER to 192.168.1.255
03[NET] received packet: from <client-ipv6>[4500] to <server-ipv6>[4500]
03[NET] waiting for data on sockets
13[MGR] ignoring request with ID 1, already processing
03[NET] received packet: from <client-ipv6>[4500] to <server-ipv6>[4500]
03[NET] waiting for data on sockets
11[MGR] ignoring request with ID 1, already processing
14[CFG] sending DHCP DISCOVER to 192.168.1.255
05[CFG] received DHCP OFFER 192.168.1.116 from 192.168.1.1
14[CFG] sending DHCP REQUEST for 192.168.1.116 to 192.168.1.1
14[CFG] sending DHCP REQUEST for 192.168.1.116 to 192.168.1.1
14[CFG] sending DHCP REQUEST for 192.168.1.116 to 192.168.1.1
06[CFG] received DHCP ACK for 192.168.1.116
14[IKE] assigning virtual IP 192.168.1.116 to peer 'client'
14[IKE] peer requested virtual IP %any6
14[IKE] no virtual IP found for %any6 requested by 'client'
14[CFG] selected proposal: ESP:AES_GCM_16_128/NO_EXT_SEQ
14[IKE] CHILD_SA child{1} established with SPIs <spi1> <spi2> and TS 0.0.0.0/0 === 192.168.1.116/32
```

**Reply by tobiasbrunner** [reply]
Ah, sorry. These log messages are only seen on platforms other than Linux. On Linux, the plugin does not log anything (unless there is an error when creating the AF_PACKET socket or setting the ARP filter on it).

**Reply by hgl** [reply]
Thanks for letting me know.

I just did some more tests.

If I don't specify `if_id_in` and `if_id_out` in swanctl.conf, my client can access the network behind the server. However, `ip neighbor` shows that the entry corresponding to the client has `FAILED` instead of a `lladder`.

According to Wireshark, the IP packet of a DNS query sent by the client to the LAN dns server isn't wrapped in an Ethernet frame, but "Linux cooked capture v1". (tcpdump also doesn't show the DNS responses, but my client can get them, and according to TRACE rule, the responses are sent from the DNS server to my client. Not sure why tcpdump isn't able to capture such packets.)

I have two questions if you don't mind:

1. Why my client's IP packets doesn't need to be wrapped in Ethernet frames and can be delivered when there is no XFRM interfaces involved?
2. It seems the farp plugin never worked on my device? I wonder if it's possible that it's somehow silently failing? The OS is vanilla OpenWrt, maybe it lacks some kernel module that is needed by the farp plugin?

**Reply by tobiasbrunner** [reply]
> 1\. Why my client's IP packets doesn't need to be wrapped in Ethernet frames

Probably just a Wireshark/Linux capture artifact. And yes, you won't be able to capture outbound plaintext traffic (see [FAQ](https://wiki.strongswan.org/projects/strongswan/wiki/FAQ#Capturing-outbound-plaintext-packets-with-tcpdumpwireshark)).

> and can be delivered when there is no XFRM interfaces involved?

Maybe there was a routing issue with the XFRM interface (e.g. related to _rp_filter_).

> 2\. It seems the farp plugin never worked on my device?

It does seem to work now, no? (Otherwise, how would traffic get routed back to your client?)

**Reply by hgl** [reply]
> how would traffic get routed back to your client?

But if ARP is working, shouldn't `ip neighbor` display a `lladdr` instead of `FAILED`?

> Maybe there was a routing issue with the XFRM interface (e.g. related to rp_filter).

I just enabled `if_id_*` again, my XFRM's IPv4 rp_filter is `0`, so it's probably something else, and it behaved exactly as my previous Wireshark screenshots showed. I don't understand two things though:

1. I tried giving the XFRM interface a LAN IP like 192.168.1.2/32 or a different IP range like 192.168.2.1/24 or not giving any, the ARP requests only show up in the LAN interface. This seems understandable, since the routing table says any devices of 192.168.1.0/24 should belong to the LAN interface. So XFRM's routing issue (if there is any) shouldn't be at play here?
2. If I'm not wrong, my client first sends a DNS query to 192.168.1.1 on the XFRM interface, so kernel asks for the client's MAC address on LAN. Could its not getting any ARP responses on LAN prove that at least in the XFRM case, farp is not listening on the LAN interface?

**Reply by tobiasbrunner** [reply]
> But if ARP is working, shouldn't `ip neighbor` display a `lladdr` instead of `FAILED`?

Possibly. On what host did you try that? Did you capture ARP traffic too? (Plus whatever traffic actually is working even without proper ARP.)

> I just enabled `if_id_*` again, my XFRM's IPv4 rp_filter is `0`, so it's probably something else

There is a global setting too, not just one per interface (`all`).

> So XFRM's routing issue (if there is any) shouldn't be at play here?

Probably depends on what routes you have configured via XFRM interface (note that strongSwan doesn't install any, see [RouteBasedVPN](https://wiki.strongswan.org/projects/strongswan/wiki/RouteBasedVPN#XFRM-Interfaces-on-Linux) for details).

> If I'm not wrong, my client first sends a DNS query to 192.168.1.1 on the XFRM interface, so kernel asks for the client's MAC address on LAN.

Is `192.168.1.1` on the VPN gateway or on a different host?

> farp is not listening on the LAN interface?

It does not listen on specific interfaces, it uses an AF_PACKET socket (with ARP-specific packet filter), so it receives pretty much any packets.

**Reply by hgl** [reply]
Thanks for keeping helping me.

> On what host did you try that

On the IKEv2 responder (OpenWrt box).

> Did you capture ARP traffic too (Plus whatever traffic actually is working even without proper ARP.)

I tested it again. With no XFRM interface and virtual IP from DHCP, three ARP requests for my client's MAC were captured, but not any responses. A client's TCP packet I spot-checked was wrapped in an Ethernet frame, whose MAC doesn't correspond to any interface my server, so I guess farp was working but the responses were not captured by tcpdump? And `ip neigh` did show `FAILED` for my client.

> There is a global setting too

Both `sysctl net.ipv4.conf.all.rp_filter` and `sysctl net.ipv4.conf.vpn.rp_filter` show `0`.

> Is 192.168.1.1 on the VPN gateway or on a different host?

The OpenWrt box acts as the IKEv2 responder. It has three interfaces of concern:

1. br-lan 192.168.1.0/24
2. pppoe-wan
3. vpn (XFRM)

My mental model of how they interact is like this:

My client first sends a DNS query on vpn to 192.168.1.1 on br-lan, and the kernel asks for my client's MAC, by broadcasting on br-lan, which it somehow couldn't get any responses.

> It does not listen on specific interfaces

Yes, after checking farp's source code, I realized that too. I also tried to debug farp on Linux, but after this patch on 5.9.4,
```
--- a/src/libcharon/plugins/farp/farp_spoofer.c
+++ b/src/libcharon/plugins/farp/farp_spoofer.c
@@ -154,12 +154,20 @@ static bool receive_arp(private_farp_spo
 				   (struct sockaddr*)&addr, &addr_len);
 	if (len == sizeof(arp))
 	{
+		DBG1(DBG_NET, "received ARP packet from %H(%#B) to %H(%#B)",
+			chunk_create((char*)&arp.sender_ip, 4), chunk_create((char*)&arp.sender_mac, 6),
+			chunk_create((char*)&arp.target_ip, 4), chunk_create((char*)&arp.target_mac, 6));
+
 		local = host_create_from_chunk(AF_INET,
 									chunk_create((char*)&arp.sender_ip, 4), 0);
 		remote = host_create_from_chunk(AF_INET,
 									chunk_create((char*)&arp.target_ip, 4), 0);
 		if (this->listener->has_tunnel(this->listener, local, remote))
 		{
+			DBG1(DBG_NET, "send ARP packet from %H(%#B) to %H(%#B)",
+				chunk_create((char*)&arp.target_ip, 4), chunk_create((char*)&arp.target_mac, 6),
+				chunk_create((char*)&arp.sender_ip, 4), chunk_create((char*)&arp.sender_mac, 6));
+
 			send_arp(this, &arp, &addr);
 		}
 		local->destroy(local);

```

charon automatically closes itself pretty quickly after being started, I wonder if it's because logging is not supported in the `receive_arp` callback? (I'm also not sure if that's the correct way in strongswan to print IPs and Macs, please let me know.)
```
00[DMN] swanctl: successfully loaded 2 connections, 0 unloaded
04[DMN] thread 4 received 11
04[LIB] no support for capturing backtraces
no support for capturing backtraces
04[DMN] killing ourself, received critical signal
Aborted
```

**Reply by hgl** [reply]
I start to doubt if AF_PACKET socket with ARP-specific packet filter really works for XFRM interfaces. https://github.com/strongswan/strongswan/blob/1321fdb8aaaca1719c0b40d99ad10f3aa10fc7f0/src/libcharon/plugins/farp/farp_spoofer.c#L210

**Reply by tobiasbrunner** [reply]
> With no XFRM interface and virtual IP from DHCP, three ARP requests for my client's MAC were captured, but not any responses. A client's TCP packet I spot-checked was wrapped in an Ethernet frame, whose MAC doesn't correspond to any interface my server, so I guess farp was working but the responses were not captured by tcpdump?

ARP packets generated by the _farp_ plugin should have the MAC address of the interface on which the ARP packet was received. So Ethernet frames from hosts behind the gateway should have a valid MAC address in the header.




> charon automatically closes itself pretty quickly after being started, I wonder if it's because logging is not supported in the `receive_arp` callback?

No, you used the `%H`and `%#B` printf-specifiers incorrectly. The first expects a `host_t*` (which you haven't passed at all, pass `local` and `remote` after creating those) and the second expects a `chunk_t*`, not a `chunk_t` directly (so you have to store the chunks in variables and then pass a pointer to them in the DBG statements).

**Reply by hgl** [reply]
> No, you used the %Hand %#B printf-specifiers incorrectly.

Thanks for the heads up.

After using this patch, I managed to see the logs:

```
--- a/src/libcharon/plugins/farp/farp_spoofer.c
+++ b/src/libcharon/plugins/farp/farp_spoofer.c
@@ -149,6 +149,8 @@ static bool receive_arp(private_farp_spo
 	arp_t arp;
 	ssize_t len;
 	host_t *local, *remote;
+	chunk_t smac;
+	chunk_t tmac;

 	len = recvfrom(this->skt, &arp, sizeof(arp), MSG_DONTWAIT,
 				   (struct sockaddr*)&addr, &addr_len);
@@ -158,8 +160,19 @@ static bool receive_arp(private_farp_spo
 									chunk_create((char*)&arp.sender_ip, 4), 0);
 		remote = host_create_from_chunk(AF_INET,
 									chunk_create((char*)&arp.target_ip, 4), 0);
+
+		smac = chunk_create((char*)&arp.sender_mac, 6);
+		tmac = chunk_create((char*)&arp.target_mac, 6);
+		DBG1(DBG_NET, "received ARP packet from %H(%#B) to %H(%#B)",
+			local, &smac,
+			remote, &tmac);
+
 		if (this->listener->has_tunnel(this->listener, local, remote))
 		{
+			DBG1(DBG_NET, "respond ARP packet to %H(%#B) from %H(%#B)",
+				local, &smac,
+				remote, &tmac);
+
 			send_arp(this, &arp, &addr);
 		}
 		local->destroy(local);
```

And I found out something unexpected:

From the logs, it seems charon only sees ARP requests from hosts (192.168.1.0/24) in LAN to my OpenWrt box (192.168.1.1/24) (asking for 192.168.1.1's MAC address), or to other hosts in LAN (192.168.1.0/24). It never sees requests from the OpenWrt box itself (192.168.1.1/24) to LAN hosts (including the VPN client in question). So it couldn't answer kernel's requests for my VPN client's MAC address.

It does answer to hosts in LAN asking for the client's MAC address, although my patched log doesn't show what MAC address it responses with.

I think this is by design? So farp can't make the machines it runs on to know the MAC address of a roadwarrior? Maybe my understanding of the whole thing is wrong?

> should have the MAC address of the interface on which the ARP packet was received

I previously said the TCP packet I spot-checked was wrapped in an Ethernet frame, whose MAC address didn't correspond to any interfaces. Upon closer checking, I found out it actually corresponded to client's own physical MAC. The client is an iPhone, upon its native IKEv2 VPN connected to charon on OpenWrt, it made TCP connections with its own WiFi MAC. I don't understand how that MAC relates to the MAC charon responded on its behalf to LAN hosts (which should be OpenWrt's LAN interface's MAC). Would appreciate it if you could shed some light. I guess some of my understandings of how ARP works must be incorrect.

**Reply by tobiasbrunner** [reply]
> It does answer to hosts in LAN asking for the client's MAC address, although my patched log doesn't show what MAC address it responses with.

That's determined in `send_arp()`.

> I think this is by design? So farp can't make the machines it runs on to know the MAC address of a roadwarrior?

Is an ARP request actually sent? If there is a /32 route for the virtual IP, there probably isn't, so make sure such a route exists (again, strongSwan installs them automatically only if you **don't** use XFRM interfaces).

> Upon closer checking, I found out it actually corresponded to client's own physical MAC.

Was that a decrypted packet captured on the VPN gateway? If so, that's probably normal as the Ethernet header of the original ESP packet could get reported in tcpdump/Wireshark also for the decrypted packet. If you captured it on a host in the LAN, then you should never see the MAC address of the actual client (at least not if it's not in that LAN itself).

**Reply by hgl** [reply]
> so make sure such a route exists

This nailed it! After manually adding a /32 route to my client via the vpn interface, my client is able to access the Internet after connecting to the VPN gateway. Thank you so much!

And I think I have a better understanding now. The farp plugin is only needed for hosts in the remote traffic selectors to be able to talk to VPN clients, the VPN gateway itself doesn't rely on ARP to talk the clients.

> strongSwan installs them automatically only if you don't use XFRM interfaces

I still don't understand this part.

When no XFRM interfaces are used, I don't see any /32 route to the client according to `ip route`. At least not with this swanctl.conf (but the client is able to access the Internet):

```
connections {
	home {
		version = 2
		pools = dhcp
		proposals = aes128gcm16-prfsha256-ecp256,aes128gcm16-prfsha256-x25519
		send_certreq = no
		local {
			auth = pubkey
		}
		remote {
			auth = pubkey
		}
		children {
			child {
				local_ts = 0.0.0.0/0
				esp_proposals = aes128gcm16-ecp256,aes128gcm16-x25519
			}
		}
	}
}
```

Why in the XFRM case, an explicit route is required?

> the Ethernet header of the original ESP packet could get reported in tcpdump/Wireshark also for the decrypted packet

It seems tcpdump doesn't support IPsec very well? I wonder if it's a limitation of the Linux kernel or tcpdump itself. Will I have a more accurate picture with tools like Wireshark's dumpecap. (It's not straightforward to compile dumpcap for OpenWrt, so I haven't had a chance to test it out.)

**Reply by tobiasbrunner** [reply]
> When no XFRM interfaces are used, I don't see any /32 route to the client according to `ip route`.

They are installed in table 220, by default. So check `ip route list table 220`.

> I wonder if it's a limitation of the Linux kernel or tcpdump itself.

Kernel (also see the FAQ entry I linked above for other limitations).

**Reply by hgl** [reply]
> They are installed in table 220

Found them!

Do you think it'd be hard to make auto-route-insertion work for the XFRM case? I'm not sure if it's possible to find  a `if_id`'s corresponding interface?

**Reply by tobiasbrunner** [reply]
Yes, doing that automatically is tricky, because there are multiple ways to create the devices (before even starting the IKE daemon, or dynamically either per IKE or per CHILD_SA e.g. in a vici or updown script). If you know the name of the interface, you can install routes in a custom updown script (the peer's virtual IP is passed to it). Alternatively, you could reserve a subnet out of the LAN subnet for your virtual IPs, then either use static DHCP assignment for the VPN clients or exclude the subnet in the DHCP daemon and assign virtual IPs via configured IP pool, then you can install the route when you create the XFRM interface (assuming you use one interface for all clients).

**Reply by hgl** [reply]
Thanks for the advice. This means I actually no longer need the DHCP plugin. Working like a charm.

**Reply by hgl** [reply]
Hope I could mark one of your replies as the answer.
